src/zenml/steps/base_step.py

```diff
@@ -291,37 +291,87 @@ def _prepare_parameter_spec(self) -> None:
                     f"json serializable parameter values."
                 ) from e
 
+    def _prepare_input_artifacts(
+        self, *artifacts: Channel, **kw_artifacts: Channel
+    ) -> Dict[str, Channel]:
+        """Verifies and prepares the input artifacts for running this step.
+
+        Args:
+            *artifacts: Positional input artifacts passed to
+                the __call__ method.
+            **kw_artifacts: Keyword input artifacts passed to
+                the __call__ method.
+
+        Returns:
+            Dictionary containing both the positional and keyword input
+            artifacts.
+
+        Raises:
+            StepInterfaceError: If there are too many or too few artifacts.
+        """
+        input_artifact_keys = list(self.INPUT_SPEC.keys())
+        if len(artifacts) > len(input_artifact_keys):
+            raise StepInterfaceError(
+                f"Too many input artifacts for step '{self.step_name}'. "
+                f"This step expects {len(input_artifact_keys)} artifact(s) "
+                f"but got {len(artifacts) + len(kw_artifacts)}."
+            )
+
+        combined_artifacts = {}
+
+        for i, artifact in enumerate(artifacts):
+            key = input_artifact_keys[i]
+            combined_artifacts[key] = artifact
+
+        for key, artifact in kw_artifacts.items():
+            if key in combined_artifacts:
+                # an artifact for this key was already set by
+                # the positional input artifacts
+                raise StepInterfaceError(
+                    f"Unexpected keyword argument '{key}' for step "
+                    f"'{self.step_name}'. An artifact for this key was "
+                    f"already passed as a positional argument."
+                )
+
+            combined_artifacts[key] = artifact
+
+        # check if there are any missing or unexpected artifacts
+        expected_artifacts = set(self.INPUT_SPEC.keys())
+        actual_artifacts = set(combined_artifacts.keys())
+        missing_artifacts = expected_artifacts - actual_artifacts
+        unexpected_artifacts = actual_artifacts - expected_artifacts
+
+        if missing_artifacts:
+            raise StepInterfaceError(
+                f"Missing input artifacts for step "
+                f"'{self.step_name}': {missing_artifacts}."
+            )
+
+        if unexpected_artifacts:
+            raise StepInterfaceError(
+                f"Got unexpected input artifacts for step "
+                f"'{self.step_name}': {unexpected_artifacts}."
+            )
+
+        return combined_artifacts
+
     def __call__(
-        self, **artifacts: BaseArtifact
+        self, *artifacts: Channel, **kw_artifacts: Channel
```

Comment:
Do we need to expose the tfx-layer here? I liked that it was ZenML BaseArtifact. What would be the motivation for the switch? :-)

Reply:
I would have liked to keep it a BaseArtifact, but I was debugging and realized it's actually channels. I (wrongly) assumed that these were BaseArtifacts when doing type annotations in the first place I think

Reply:
But wouldnt this be a bit weird when writing the `connect` function for the user? Or do our BaseStep `__call__` functions all returns Channels?

Reply:
Yeah they all return channels as well, so it should now be consistent at least

Reply:
Should we consider returning artifacts?

Reply:
Would definitely be nicer I think, we'd just have to connect BaseArtifacts to a Channel somehow. Can we add custom properties to a BaseArtifact or does that somehow mess with tfx?

Reply:
We have to investigate but its linked to the same logic for adding custom properties to the metadata store which keeps popping up. It also is a bit logically messed up as Channels point to artifacts in the TFX design

Reply:
Yeah that's a little strange then, I added a TODO for now